tensorflow_io/avro/20190307-avro-dataset.md

```diff
@@ -0,0 +1,299 @@
+
+Loading Avro formatted data will enable the data science community to leverage TensorFlow more easily. Currently, loading Avro formatted data into TensorFlow is not supported. Thus, ways of getting data into TensorFlow requires a conversion to other data formats, e.g. protobuf, CSV, or proprietary formats. Such custom solutions are difficult to maintain and do not generalize well for other uses.
+
+## Design Proposal
+
+The Avro dataset shall support
+
+- Transform primitive and nested Avro types into TensorFlow sparse and dense tensors
+- The creation of dimensions by (see also examples in Tables 2 and 3)
+  - Combining of several rows -- also called batching
+  - Combining several Avro fields to define indices and values for sparse tensors
+  - Combining and reshaping of elements of an Avro non-nested array into a multi-dimensional tensor using row-major-order
+  - Nestings of Avro typed arrays
+- Filling in scalars or tensors for missing values whenever the number of provided elements does not match the desired number of elements by using default values -- if no defaults or the wrong defaults are provided this results in a runtime error surfaced through an exception in python and erroneous status returned in c/c++
+- Full support for N-dimensional dense tensors and N-dimensional sparse tensors
+- A mapping of Avro primitive types to TensorFlow primitive types (see Table 1)
+- For Avro&#39;s null type the default shall be assumed if provided; otherwise fail
+- An API/syntax that follows the one in Spark/Avro tooling that allows to resolve
+  - Elements in an array
+  - Elements in a map
+  - Nested records
+  - Namespaces
+  - Branches in unions -- will resolve them but will use strict typing which will render them useless whenever an attribute has multiple types or whenever a null is encountered. Note, TensorFlow uses strict typing for it&#39;s tensors.
+- An API for filtering in arrays
+- The mapping of Avro enumerations to TensorFlow strings carrying the symbol name
+- The mapping of Avro fixed types to TensorFlow strings
+- The same operating systems as supported by TensorFlow -- Note, that this only accounts for the development of the avro dataset and not it&#39;s dependencies
+- Python 2.x and 3.x following TensorFlow&#39;s requirements
+- Thread safety; Note that TensorFlow&#39;s  dataset API may use multiple threads
+
+| **Avro primitive type** | **TensorFlow primitive type** |
+| --- | --- |
+| null: no value | If provided default is assumed otherwise fail |
+| boolean: a binary value | tf.bool |
+| int: 32-bit signed integer | tf.int32 |
+| long: 64-bit signed integer | tf.in64 |
+| float: single precision 32-bit IEEE floating point number | tf.float32 |
+| double: double precision 64-bit IEEE floating point number | tf.float64 |
+| fixed: a fixed length byte sequence | tf.string |
+| bytes: a sequence of 8-bit unsigned bytes | tf.string |
+| string: unicode character sequence | tf.string |
+| enum: enumeration | tf.string |
```

Comment:
Noted that Avro's enum is string:
https://avro.apache.org/docs/1.8.1/spec.html#Enums

So it is not really C/C++ types of enum (int) 👍

Reply:
Perfect. Thanks.